File: yunshu_music/ios/Runner/MusicPlayDataService.swift
import Foundation

class MusicPlayDataService {
    
    public static let INSTANCE = MusicPlayDataService()
    
    private let NOW_PLAY_MEDIA_ID_KEY = "NOW_PLAY_MEDIA_ID_KEY"
    private let PLAY_MODE_KEY = "PLAY_MODE_KEY"
    private let PLAY_LIST_KEY = "PLAY_LIST_KEY"
    
    private var MUSIC_LIST:[Music] = []
    private var PLAY_LIST:[Music] = []
    private var RANDOM_SET = Set<Music>()
    
    private var nowPlayIndex:Int = -1
    
    private var nowPlayMusic:Music?
    
    private var playMode:MusicPlayMode
    
    private let kv:UserDefaults
    
    init() {
        self.nowPlayIndex = -1
        self.kv = UserDefaults.standard
        if kv.object(forKey: PLAY_MODE_KEY) == nil {
            kv.set(MusicPlayMode.SEQUENCE.rawValue, forKey:PLAY_MODE_KEY)
            self.playMode = MusicPlayMode.SEQUENCE
        } else {
            let mode:String = kv.string(forKey: PLAY_MODE_KEY) ?? MusicPlayMode.SEQUENCE.rawValue
            self.playMode = MusicPlayMode.valueOf(mode)
        }
    }
    
    func getNowPlayIndex() -> Int {
        return nowPlayIndex
    }
    
    func getNowPlayMusic() -> Music? {
        return nowPlayMusic
    }
    
    func getPlayMode() -> MusicPlayMode {
        return playMode
    }
    
    func getPlayList() -> [Music] {
        return PLAY_LIST
    }
    
    func delPlayListByMediaId(mediaId:String) -> Void {
        PLAY_LIST.removeAll(where: { $0.musicId == mediaId })
        
        do{
            let jsonEncoder = JSONEncoder()
            let jsonData = try jsonEncoder.encode(PLAY_LIST.map{ $0.musicId })
            let playListString = String(data: jsonData, encoding: String.Encoding.utf8)
            kv.set(playListString, forKey: PLAY_LIST_KEY)
        }catch let error{
            print("catch error when set play list \(error)")
        }
    }
    
    func clearPlayList() -> Void {
        PLAY_LIST.removeAll()
        nowPlayIndex = -1
        kv.removeObject(forKey: PLAY_LIST_KEY)
    }
    
    func setPlayMode(playMode:MusicPlayMode) -> Void {
        self.playMode = playMode
        kv.set(playMode.rawValue, forKey: PLAY_MODE_KEY)
    }
    
    func addMusic(musicList:[Music]) -> Void {
        MUSIC_LIST.append(contentsOf: musicList)
        
        do{
            let playListString = kv.string(forKey: PLAY_LIST_KEY) ?? "[]"
            let jsonDecoder = JSONDecoder()
            let playListMusicIdList:[String] = try jsonDecoder.decode([String].self, from: playListString.data(using: String.Encoding.utf8)!)
            var playList:[Music] = []
            for mediaId in playListMusicIdList{
                if let i = MUSIC_LIST.first(where: { $0.musicId == mediaId }){
                    playList.append(i)
                }
            }
            PLAY_LIST.append(contentsOf: playList)
            
        }catch let error{
            print("catch error when add music \(error)")
        }
        
        let nowPlayMediaId = kv.string(forKey: NOW_PLAY_MEDIA_ID_KEY)
        if nowPlayMediaId != nil{
            if let i = PLAY_LIST.firstIndex(where: {$0.musicId==nowPlayMediaId}){
                nowPlayIndex = i
                nowPlayMusic = PLAY_LIST[i]
            }
        }
        if -1 == nowPlayIndex{
            self.next(userTrigger: false)
        }
    }
    
    func removeMusic(music:Music) -> Void {
        MUSIC_LIST.removeAll{ $0 == music }
    }
    
    func playFromMediaId(mediaId:String) -> Void {
        nowPlayIndex = -1;
        nowPlayMusic = nil;
        nowPlayMusic = MUSIC_LIST.first{$0.musicId == mediaId}
        if nowPlayMusic == nil{
            return
        }
        let playListIndex:Int? = PLAY_LIST.firstIndex(of: nowPlayMusic!)
        if playListIndex == nil{
            PLAY_LIST.append(nowPlayMusic!)
            nowPlayIndex = PLAY_LIST.endIndex - 1
        }else{
            nowPlayIndex = playListIndex!
        }
        
        do{
            let jsonEncoder = JSONEncoder()
            let jsonData = try jsonEncoder.encode(PLAY_LIST.map{ $0.musicId })
            let playListString = String(data: jsonData, encoding: String.Encoding.utf8)
            kv.set(playListString, forKey: PLAY_LIST_KEY)
            kv.set(nowPlayMusic!.musicId, forKey: NOW_PLAY_MEDIA_ID_KEY)
        }catch let error{
            print("catch error when set play list \(error)")
        }
    }
    
    func previous(userTrigger:Bool) -> Void {
        if nowPlayIndex - 1 < 0{
            switch playMode {
            case .RANDOMLY:
                let randomMusicListIndex = getRandom()
                nowPlayMusic = MUSIC_LIST[randomMusicListIndex]
                PLAY_LIST.removeAll{$0==nowPlayMusic}
                PLAY_LIST.append(nowPlayMusic!)
                nowPlayIndex = 0
            case .SEQUENCE:
                let sequenceMusicListIndex = toSequencePrevious()
                nowPlayMusic = MUSIC_LIST[sequenceMusicListIndex]
                PLAY_LIST.removeAll{$0==nowPlayMusic}
                PLAY_LIST.insert(nowPlayMusic!, at: 0)
                nowPlayIndex = 0;
            case .LOOP:
                if userTrigger{
                    let loopMusicListIndex = toSequencePrevious()
                    nowPlayMusic = MUSIC_LIST[loopMusicListIndex]
                    PLAY_LIST.removeAll{$0==nowPlayMusic}
                    PLAY_LIST.insert(nowPlayMusic!, at: 0)
                    nowPlayIndex = 0;
                }
            }
        }else if userTrigger || playMode != MusicPlayMode.LOOP{
            nowPlayIndex -= 1
            nowPlayMusic = PLAY_LIST[nowPlayIndex]
        }
        
        do{
            let jsonEncoder = JSONEncoder()
            let jsonData = try jsonEncoder.encode(PLAY_LIST.map{ $0.musicId })
            let playListString = String(data: jsonData, encoding: String.Encoding.utf8)
            kv.set(playListString, forKey: PLAY_LIST_KEY)
            kv.set(nowPlayMusic!.musicId, forKey: NOW_PLAY_MEDIA_ID_KEY)
        }catch let error{
            print("catch error when set play list \(error)")
        }
    }
    
    func next(userTrigger:Bool) -> Void {
        if nowPlayIndex+1 >= PLAY_LIST.endIndex{
            switch playMode {
            case .RANDOMLY:
                let randomMusicListIndex = getRandom()
                nowPlayMusic = MUSIC_LIST[randomMusicListIndex]
                PLAY_LIST.removeAll{$0==nowPlayMusic}
                PLAY_LIST.append(nowPlayMusic!)
                nowPlayIndex += 1
            case .SEQUENCE:
                let sequenceMusicListIndex = toSequenceNext()
                nowPlayMusic = MUSIC_LIST[sequenceMusicListIndex]
                PLAY_LIST.removeAll{$0==nowPlayMusic}
                PLAY_LIST.append(nowPlayMusic!)
                nowPlayIndex += 1
            case .LOOP:
                if userTrigger{
                    let loopMusicListIndex = toSequenceNext()
                    nowPlayMusic = MUSIC_LIST[loopMusicListIndex]
                    PLAY_LIST.removeAll{$0==nowPlayMusic}
                    PLAY_LIST.append(nowPlayMusic!)
                    nowPlayIndex += 1
                }
            }
        }
    }
    
    private func getRandom() -> Int {
        var canPlayList:[Music] = MUSIC_LIST.filter{!RANDOM_SET.contains($0)}.filter{!PLAY_LIST.contains($0)}
        if canPlayList.isEmpty{
            RANDOM_SET.removeAll()
            canPlayList = MUSIC_LIST
        }
        let canPlayListIndex:Int = Int.random(in: 0...canPlayList.endIndex)
        let music:Music = canPlayList[canPlayListIndex]
        RANDOM_SET.insert(music)
        return MUSIC_LIST.firstIndex(of: music)!
    }
    
    private func toSequenceNext() -> Int {
        if nowPlayIndex == -1{
            return 0
        }
        let mediaItem:Music = PLAY_LIST[nowPlayIndex]
        let musicListIndex:Int = MUSIC_LIST.firstIndex(of: mediaItem)!
        if musicListIndex+1 >= MUSIC_LIST.endIndex {
            return 0
        }else{
            return musicListIndex + 1
        }
    }
    
    private func toSequencePrevious()->Int{
        if nowPlayIndex == -1{
            return MUSIC_LIST.endIndex - 1
        }
        let mediaItem:Music = PLAY_LIST[nowPlayIndex]
        let musicListIndex:Int = MUSIC_LIST.firstIndex{$0 == mediaItem}!
        if musicListIndex - 1 < 0{
            return MUSIC_LIST.endIndex - 1
        }else{
            return musicListIndex - 1
        }
    }
    
}

File: yunshu_music/android/app/src/main/java/top/itning/yunshu_music/service/MusicPlayDataService.java
package top.itning.yunshu_music.service;

import android.net.Uri;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.media3.common.MediaItem;
import androidx.media3.common.MediaMetadata;

import com.tencent.mmkv.MMKV;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Random;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * @author itning
 * @since 2021/10/12 14:52
 */
public class MusicPlayDataService {
    private static final String NOW_PLAY_MEDIA_ID_KEY = "NOW_PLAY_MEDIA_ID_KEY";
    private static final String PLAY_MODE_KEY = "PLAY_MODE";
    private static final String PLAY_LIST_KEY = "PLAY_LIST";
    private static final String LYRICS_URI_KEY = "lyricUri";
    private final List<MediaItem> MUSIC_LIST = new ArrayList<>();
    private final List<MediaItem> PLAY_LIST = new ArrayList<>();
    private final Set<MediaItem> RANDOM_SET = new HashSet<>();
    private int nowPlayIndex;
    private MediaItem nowPlayMusic;
    private MusicPlayMode playMode;
    private final MMKV kv;

    public MusicPlayDataService() {
        this.nowPlayIndex = -1;
        kv = MMKV.defaultMMKV();
        try {
            String mode = kv.decodeString(PLAY_MODE_KEY, MusicPlayMode.SEQUENCE.name());
            this.playMode = MusicPlayMode.valueOf(mode);
        } catch (Exception e) {
            kv.encode(PLAY_MODE_KEY, MusicPlayMode.SEQUENCE.name());
            this.playMode = MusicPlayMode.SEQUENCE;
        }
    }

    public static MediaItem buildMediaItem(String mediaId, String musicUri, String name, String singer, String coverUri, String lyricUri) {
        Bundle extras = new Bundle();
        if (lyricUri != null) {
            extras.putString(LYRICS_URI_KEY, lyricUri);
        }
        MediaMetadata metadata = new MediaMetadata.Builder()
                .setTitle(name)
                .setArtist(singer)
                .setArtworkUri(coverUri == null ? null : Uri.parse(coverUri))
                .setExtras(extras)
                .build();
        MediaItem.Builder builder = new MediaItem.Builder()
                .setMediaId(mediaId)
                .setMediaMetadata(metadata);
        if (musicUri != null) {
            builder.setUri(musicUri);
        }
        return builder.build();
    }

    @Nullable
    public String getMediaUri(MediaItem item) {
        if (item.localConfiguration == null || item.localConfiguration.uri == null) {
            return null;
        }
        return item.localConfiguration.uri.toString();
    }

    @Nullable
    public CharSequence getTitle(MediaItem item) {
        return item.mediaMetadata.title;
    }

    @Nullable
    public CharSequence getSinger(MediaItem item) {
        return item.mediaMetadata.artist;
    }

    @Nullable
    public Uri getCoverUri(MediaItem item) {
        return item.mediaMetadata.artworkUri;
    }

    @Nullable
    public String getLyricUri(MediaItem item) {
        Bundle extras = item.mediaMetadata.extras;
        return extras == null ? null : extras.getString(LYRICS_URI_KEY);
    }

    public int getNowPlayIndex() {
        return nowPlayIndex;
    }

    public MediaItem getNowPlayMusic() {
        return nowPlayMusic;
    }

    public MusicPlayMode getPlayMode() {
        return playMode;
    }

    public List<MediaItem> getPlayList() {
        return PLAY_LIST;
    }

    public boolean isMusicListEmpty() {
        return MUSIC_LIST.isEmpty();
    }

    public void delPlayListByMediaId(String mediaId) {
        if (nowPlayMusic != null && mediaId.equals(nowPlayMusic.mediaId)) {
            return;
        }
        int removeIndex = -1;
        for (int i = 0; i < PLAY_LIST.size(); i++) {
            if (mediaId.equals(PLAY_LIST.get(i).mediaId)) {
                removeIndex = i;
                break;
            }
        }
        if (-1 == removeIndex) {
            return;
        }
        PLAY_LIST.remove(removeIndex);
        if (removeIndex < nowPlayIndex) {
            nowPlayIndex--;
        }
        String playListString = PLAY_LIST.stream().map(it -> it.mediaId).collect(Collectors.joining("@"));
        kv.encode(PLAY_LIST_KEY, playListString);
    }

    public void clearPlayList() {
        RANDOM_SET.clear();
        PLAY_LIST.clear();
        if (nowPlayMusic != null) {
            PLAY_LIST.add(nowPlayMusic);
            nowPlayIndex = 0;
            String playListString = PLAY_LIST.stream().map(it -> it.mediaId).collect(Collectors.joining("@"));
            kv.encode(PLAY_LIST_KEY, playListString);
        } else {
            nowPlayIndex = -1;
            kv.removeValueForKey(PLAY_LIST_KEY);
        }
    }

    public void setPlayMode(MusicPlayMode playMode) {
        this.playMode = playMode;
        RANDOM_SET.clear();
        kv.encode(PLAY_MODE_KEY, playMode.name());
    }

    public void addMusic(List<MediaItem> musicList) {
        MUSIC_LIST.clear();
        MUSIC_LIST.addAll(musicList);
        PLAY_LIST.clear();
        nowPlayIndex = -1;
        String playListString = kv.decodeString(PLAY_LIST_KEY, "");

        List<String> playListMusicIdList = Arrays.asList(playListString.split("@"));
        List<MediaItem> playList = new ArrayList<>(playListMusicIdList.size());
        for (String mediaId : playListMusicIdList) {
            MUSIC_LIST.stream().filter(it -> mediaId.equals(it.mediaId)).findFirst().ifPresent(playList::add);
        }
        PLAY_LIST.addAll(playList);

        String nowPlayMediaId = kv.decodeString(NOW_PLAY_MEDIA_ID_KEY);
        if (null != nowPlayMediaId) {
            for (int i = 0; i < PLAY_LIST.size(); i++) {
                if (nowPlayMediaId.equals(PLAY_LIST.get(i).mediaId)) {
                    nowPlayIndex = i;
                    nowPlayMusic = PLAY_LIST.get(i);
                    break;
                }
            }
        }
        if (-1 == nowPlayIndex) {
            this.next(false);
        }
    }

    public void removeMusic(MediaItem music) {
        MUSIC_LIST.remove(music);
    }

    public void playFromMediaId(String mediaId) {
        nowPlayIndex = -1;
        nowPlayMusic = null;
        for (int i = 0; i < MUSIC_LIST.size(); i++) {
            MediaItem item = MUSIC_LIST.get(i);
            if (mediaId.equals(item.mediaId)) {
                nowPlayMusic = item;
                break;
            }
        }
        if (null == nowPlayMusic) {
            return;
        }
        int playListIndex = PLAY_LIST.indexOf(nowPlayMusic);
        if (-1 == playListIndex) {
            PLAY_LIST.add(nowPlayMusic);
            nowPlayIndex = PLAY_LIST.size() - 1;
        } else {
            nowPlayIndex = playListIndex;
        }
        String playListString = PLAY_LIST.stream().map(it -> it.mediaId).collect(Collectors.joining("@"));
        kv.encode(PLAY_LIST_KEY, playListString);
        kv.encode(NOW_PLAY_MEDIA_ID_KEY, nowPlayMusic.mediaId);
    }

    public void previous(boolean userTrigger) {
        if (MUSIC_LIST.isEmpty()) {
            return;
        }
        if (nowPlayIndex - 1 < 0) {
            switch (playMode) {
                case RANDOMLY:
                    int randomMusicListIndex = getRandom();
                    nowPlayMusic = MUSIC_LIST.get(randomMusicListIndex);
                    PLAY_LIST.remove(nowPlayMusic);
                    PLAY_LIST.add(0, nowPlayMusic);
                    nowPlayIndex = 0;
                    break;
                case SEQUENCE:
                    int sequenceMusicListIndex = toSequencePrevious();
                    nowPlayMusic = MUSIC_LIST.get(sequenceMusicListIndex);
                    PLAY_LIST.remove(nowPlayMusic);
                    PLAY_LIST.add(0, nowPlayMusic);
                    nowPlayIndex = 0;
                    break;
                case LOOP:
                    if (userTrigger) {
                        int loopMusicListIndex = toSequencePrevious();
                        nowPlayMusic = MUSIC_LIST.get(loopMusicListIndex);
                        PLAY_LIST.remove(nowPlayMusic);
                        PLAY_LIST.add(0, nowPlayMusic);
                        nowPlayIndex = 0;
                    }
                    break;
            }
        } else if (userTrigger || playMode != MusicPlayMode.LOOP) {
            nowPlayIndex--;
            nowPlayMusic = PLAY_LIST.get(nowPlayIndex);
        }
        String playListString = PLAY_LIST.stream().map(it -> it.mediaId).collect(Collectors.joining("@"));
        kv.encode(PLAY_LIST_KEY, playListString);
        kv.encode(NOW_PLAY_MEDIA_ID_KEY, nowPlayMusic.mediaId);
    }

    public void next(boolean userTrigger) {
        if (MUSIC_LIST.isEmpty()) {
            return;
        }
        if (nowPlayIndex + 1 >= PLAY_LIST.size()) {
            switch (playMode) {
                case RANDOMLY:
                    int randomMusicListIndex = getRandom();
                    nowPlayMusic = MUSIC_LIST.get(randomMusicListIndex);
                    PLAY_LIST.remove(nowPlayMusic);
                    PLAY_LIST.add(nowPlayMusic);
                    nowPlayIndex = PLAY_LIST.size() - 1;
                    break;
                case SEQUENCE:
                    int sequenceMusicListIndex = toSequenceNext();
                    nowPlayMusic = MUSIC_LIST.get(sequenceMusicListIndex);
                    PLAY_LIST.remove(nowPlayMusic);
                    PLAY_LIST.add(nowPlayMusic);
                    nowPlayIndex = PLAY_LIST.size() - 1;
                    break;
                case LOOP:
                    if (userTrigger) {
                        int loopMusicListIndex = toSequenceNext();
                        nowPlayMusic = MUSIC_LIST.get(loopMusicListIndex);
                        PLAY_LIST.remove(nowPlayMusic);
                        PLAY_LIST.add(nowPlayMusic);
                        nowPlayIndex = PLAY_LIST.size() - 1;
                    }
                    break;
            }
        } else if (userTrigger || playMode != MusicPlayMode.LOOP) {
            nowPlayIndex++;
            nowPlayMusic = PLAY_LIST.get(nowPlayIndex);
        }
        String playListString = PLAY_LIST.stream().map(it -> it.mediaId).collect(Collectors.joining("@"));
        kv.encode(PLAY_LIST_KEY, playListString);
        kv.encode(NOW_PLAY_MEDIA_ID_KEY, nowPlayMusic.mediaId);
    }

    private int getRandom() {
        List<MediaItem> canPlayList = MUSIC_LIST.stream()
                .filter(item -> !RANDOM_SET.contains(item))
                .filter(item -> !PLAY_LIST.contains(item))
                .collect(Collectors.toList());
        if (canPlayList.isEmpty()) {
            RANDOM_SET.clear();
            canPlayList = MUSIC_LIST.stream()
                    .filter(item -> !item.equals(nowPlayMusic))
                    .collect(Collectors.toList());
        }
        if (canPlayList.isEmpty()) {
            canPlayList = MUSIC_LIST;
        }
        Random random = new Random();
        int canPlayListIndex = random.nextInt(canPlayList.size());
        MediaItem mediaItem = canPlayList.get(canPlayListIndex);
        RANDOM_SET.add(mediaItem);
        return MUSIC_LIST.indexOf(mediaItem);
    }

    private int toSequenceNext() {
        if (nowPlayIndex == -1) {
            return 0;
        }
        MediaItem mediaItem = PLAY_LIST.get(nowPlayIndex);
        int musicListIndex = MUSIC_LIST.indexOf(mediaItem);
        if (musicListIndex + 1 >= MUSIC_LIST.size()) {
            return 0;
        } else {
            return musicListIndex + 1;
        }
    }

    private int toSequencePrevious() {
        if (nowPlayIndex == -1) {
            return MUSIC_LIST.size() - 1;
        }
        MediaItem mediaItem = PLAY_LIST.get(nowPlayIndex);
        int musicListIndex = MUSIC_LIST.indexOf(mediaItem);
        if (musicListIndex - 1 < 0) {
            return MUSIC_LIST.size() - 1;
        } else {
            return musicListIndex - 1;
        }
    }
}
